Revamp AWS quick-start documentation to include Redshift + Databricks (#494)
snowplow/documentation: docs/getting-started-on-snowplow-open-source/quick-start-aws/index.md

    - [Configured](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-config) against a role that has the AdminstratorAccess policy attached
        - AdminstratorAccess allows all actions on all AWS services and shouldn't be used in production
- [Terraform 1.0.0](https://www.terraform.io/downloads.html) or higher installed
    - Follow the instructions to make sure the `terraform` binary is available on your `PATH`. You can also use [tfenv](https://github.com/tfutils/tfenv) to help manage your Terraform installation

> - Download [the latest igluctl](/docs/pipeline-components-and-applications/iglu/igluctl-2/index.md) which allows you to publish schemas for your [custom events](/docs/understanding-tracking-design/out-of-the-box-vs-custom-events-and-entities/index.md#custom-events) and [entities](/docs/understanding-tracking-design/predefined-vs-custom-entities/index.md#custom-contexts) to [Iglu (your schema registry)](/docs/pipeline-components-and-applications/iglu/index.md)

Reviewer: Would remove this from here, I think it’s easier to deploy first and then figure out the usage. Same way as we are not asking to have `psql` installed at this point.

> The main difference is around the [VPC](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html) that the components are deployed within. In `default` you will deploy everything into a public subnet, this is the easiest route if you want to try out Snowplow as you can use [your default VPC](https://docs.aws.amazon.com/vpc/latest/userguide/default-vpc.html#view-default-vpc). However, to increase the security of your components, it is recommended and best practise to deploy components into private subnets. This ensures they are not available publicly. To use the `` `secure` `` configuration you will need [your own VPC with public and private subnets](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Scenario2.html). You can follow [this guide](https://docs.aws.amazon.com/vpc/latest/userguide/working-with-vpcs.html) for steps on how to create VPCs and subnets on AWS.

Reviewer: I think I would make this into tabs. Feel free to ignore for now, I can do that myself later

:::note

> If opting for the `secure` deployment you must ensure your private subnet layer has full access to the internet via one of NAT Gateways, NAT Instances or a Transit Gateway.  Without internet access the deployment will fail.

Reviewer: ```suggestion
If opting for the `secure` deployment, you must ensure your private subnet layer has full access to the internet via one of NAT Gateways, NAT Instances or a Transit Gateway. Without internet access the deployment will fail.
```

:::

#### Storage Options

> There are also five different storage options for you to select. The steps below will guide you through how to set each option up - its generally recommended to only load data into a single destination but it is entirely possible to deploy a pipeline to load into every available storage option for trial purposes.

Reviewer: ```suggestion
There are also five different storage options for you to select: Postgres, Redshift, Snowflake, Databricks, ???. The steps below will guide you through how to set each option up. We recommend to only load data into a single destination, but nothing prevents you from loading into multiple destinations with the same pipeline (e.g. for testing purposes).
```

### Setting up your Iglu Server

> The first step is to set up your [Iglu](/docs/pipeline-components-and-applications/iglu/index.md) Server stack.  This will mean that you can create and evolve your own [custom events & entities](/docs/understanding-tracking-design/out-of-the-box-vs-custom-events-and-entities/index.md#custom-events). Iglu enables you to store the schemas for your events & entities and fetch them as your events are being processed by your pipeline.

Reviewer: ```suggestion
The first step is to set up your [Iglu](/docs/pipeline-components-and-applications/iglu/index.md) Server stack.  This will mean that you can create and evolve your own [custom events & entities](/docs/understanding-tracking-design/out-of-the-box-vs-custom-events-and-entities/index.md#custom-events). Iglu enables you to store the schemas for your events & entities and fetch them as your events are processed by your pipeline.
```

terraform apply
```

> The deployment will **take roughly ~15 minutes to deploy** and once done will output your `iglu_server_dns_name`. Make a note of this, you'll need it when setting up your pipeline. If you have attached a custom SSL certificate and set up your own DNS records then you don't need this value.

Reviewer: ```suggestion
The deployment will **take roughly 15 minutes** and once done will output your `iglu_server_dns_name`. Make a note of this, you'll need it when setting up your pipeline. If you have attached a custom SSL certificate and set up your own DNS records, then you don't need this value.
```

> **Step 3 (optional): Add your schemas to Iglu Server**

Reviewer: I would remove this from here. I don’t think anyone would write up schemas before deploying a pipeline. And it’s distracting when we mix installation with usage. We’ll point people to the usage doc later in the flow. (Also, it’s not a good idea to duplicate these instructions across the docs — we should have a [single place](https://docs.snowplow.io/docs/understanding-tracking-design/managing-your-data-structures/iglu/) where this is explained and just link to it.)

```bash
git clone https://github.com/snowplow/quickstart-examples.git
cd quickstart-examples/terraform/aws/pipeline/default # or secure
nano terraform.tfvars # or other text editor of your choosing
```

To update your input variables, you'll need to know a couple of things:

- Your IP Address. [Help](https://duckduckgo.com/?q=ip+address&t=ffab&ia=answer).
- Your Iglu Servers DNS from [Setting up your Iglu Server.](#iglu-setup)
- Your UUIDv4 for your Iglu Servers API Key. [Help](https://duckduckgo.com/?t=ffab&q=uuid&ia=answer).
- The default VPC you will deploy your Pipeline into. [Help](https://docs.aws.amazon.com/vpc/latest/userguide/default-vpc.html#view-default-vpc).
- (optional) An SSL certificate so that your collector can receive HTTPS traffic on a domain you own. You can create one using [Amazon Certificate Manager](https://aws.amazon.com/certificate-manager/). Once you have done that, add the certificate in the `ssl_certificate_arn` variable.

> - How to generate a SSH Key.

Reviewer: ```suggestion
- How to generate an SSH Key.
```

    - On most systems you can generate a SSH Key with: `ssh-keygen -t rsa -b 4096`
    - This will output where you public key is stored, for example: `~/.ssh/id_rsa.pub`
    - You can get the value with `cat ~/.ssh/id_rsa.pub`

As mentioned above, there are several options for the pipeline's destination database. These are Postgres, Snowflake, Redshift and Databricks. For each destination there is a dedicated block of variables that will need to be filled and the `<destination>_enabled` variable set to `true`.

##### Postgres

> If you choose Postgres as destination, there are no additional steps. Respective variables need to be filled according to the desired setup. Necessary resources like a PostgreSQL RDS instance, database, table, user will be created by the Terraform modules.

Reviewer: Could you list the “respective variables” or at least clarify that those are `postgres_...` ? Also, I would make it clear that `postgres_enabled` should be set to `true` here.

> It is strongly recommended to:

Reviewer: ```suggestion
:::caution

We strongly recommend to:
```

> 1. Update the `postgres_db_password` to a new value that only **you** know
> 2. Set the `postgres_db_ip_allowlist` to a list of CIDR addresses that will need to access the database - this can be systems like BI Tools, or your home address so you can query the database from your laptop

Reviewer: ```suggestion
1. Update the `postgres_db_password` to a new value that only **you** know
2. Set the `postgres_db_ip_allowlist` to a list of CIDR addresses that will need to access the database - this can be systems like BI Tools, or your local IP address so that you can query the database from your laptop

:::
```

Do we have an example format for the allowlist somewhere?

##### Snowflake, Redshift & Databricks

> If you choose Snowflake, Redshift or Databricks as a destination, there is one more additional step. Pipeline Terraform module doesn't create necessary resources for these destinations. Unlike the Postgres resources it only deploys the loader itself. Therefore destination resources need to be created *before* deploying the pipeline.

Reviewer: ```suggestion
If you choose Snowflake, Redshift or Databricks as a destination, there is one more step. Pipeline Terraform module doesn't create the necessary resources for these destinations. Unlike with Postgres, it only deploys the loader itself. Therefore destination resources need to be created *before* deploying the pipeline.
```

The setup instructions for each destination can be found on the loader module page:

> * Snowflake: https://github.com/snowplow-devops/terraform-aws-snowflake-loader-ec2#usage

Reviewer: Not sure I like that the instructions are in a separate repo. Could we move them here?

I also have a few questions about those instructions, e.g. do we need to create the Snowflake schema? I think the latest RDB loader creates that by itself.

* Databricks: https://github.com/snowplow-devops/terraform-aws-databricks-loader-ec2#usage
* Redshift: https://github.com/snowplow-devops/terraform-aws-redshift-loader-ec2#usage

Take note of all of the inputs/outputs produced as you will need them to flesh out the `<destination>_*` variables in the `terraform.tfvars` file.

**Step 2: Run the terraform script to set up your Pipeline stack**

You can now use terraform to create your Pipeline stack. You will be asked to select a region, you can find more information about [available regions here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html).

```bash
terraform init
terraform plan
terraform apply
```

> This will output your `collector_dns_name`, `postgres_db_address`, `postgres_db_port` and `postgres_db_id`. Make a note of these, you'll need it when sending events and connecting to your database. If you have attached a custom ssl certificate and set up your own DNS records then you don't need your `collector_dns_name` as you will use your own DNS record to send events from the Snowplow trackers.

Reviewer: The postgres ones only appear if you chose postgres, right? I would clarify that.

Author: They always appear annoyingly - they are just empty values.  But yes that should be clarified.
